tempesta_fw/http_parser.c

```diff
@@ -4118,185 +4172,3193 @@ Req_Method_1CharStep: __attribute__((cold))
+			{
+				__FSM_H2_FIN(Req_HdrPsPathV, 5,
+					     TFW_TAG_HDR_H2_PATH);
+			}
+			__FSM_H2_NEXT(Req_HdrPseudo);
+		case 'a':
+			if (likely(__data_available(p, 6)
+				   && C4_INT(p + 1, 'c', 'c', 'e', 'p')
+				   && *(p + 5) == 't'))
+			{
+				__FSM_H2_FIN(Req_HdrAcceptV, 6,
+					     TFW_TAG_HDR_ACCEPT);
+			}
+			if (likely(__data_available(p, 13)
+				   && C8_INT(p + 1, 'u', 't', 'h', 'o',
+					     'r', 'i', 'z', 'a')
+				   && C4_INT(p + 9, 't', 'i', 'o', 'n')))
+			{
+				__FSM_H2_FIN(Req_HdrAuthorizationV, 13,
+					     TFW_TAG_HDR_AUTHORIZATION);
+			}
+			__FSM_H2_NEXT(Req_HdrA);
+		case 'c':
+			if (unlikely(!__data_available(p, 13)))
+				__FSM_H2_NEXT(Req_HdrC);
+			switch (PI(p + 1)) {
+			case TFW_CHAR4_INT('a', 'c', 'h', 'e'):
+				if (likely(C8_INT(p + 5, '-', 'c', 'o', 'n',
+						  't', 'r', 'o', 'l')))
+				{
+					__FSM_H2_FIN(Req_HdrCache_ControlV, 13,
+						     TFW_TAG_HDR_CACHE_CONTROL);
+				}
+				__FSM_H2_OTHER_n(5);
+			case TFW_CHAR4_INT('o', 'n', 'n', 'e'):
+				if (likely(C4_INT(p + 5, 'c', 't', 'i', 'o')
+					   && *(p + 9) == 'n'))
+				{
+					/*
+					 * Messages with connection-specific
+					 * headers must be treated as malformed
+					 * in HTTP/2 (see RFC 7540 section
+					 * 8.1.2.2 for details).
+					 */
+					__FSM_H2_DROP(Req_HdrConnection);
+				}
+				__FSM_H2_OTHER_n(5);
+			case TFW_CHAR4_INT('o', 'n', 't', 'e'):
+				if (likely(*(p + 5) == 'n'
+					   && *(p + 6) == 't'
+					   && *(p + 7) == '-'))
+				{
+					__FSM_H2_NEXT_n(Req_HdrContent_, 8);
+				}
+				__FSM_H2_OTHER_n(5);
+			case TFW_CHAR4_INT('o', 'o', 'k', 'i'):
+				if (likely(*(p + 5) == 'e'))
+				{
+					__FSM_H2_FIN(Req_HdrCookieV, 6,
+						     TFW_TAG_HDR_COOKIE);
+				}
+				__FSM_H2_OTHER_n(5);
+			default:
+				__FSM_H2_OTHER();
+			}
+		case 'h':
+			if (likely(__data_available(p, 4)
+				   && *(p + 1) == 'o'
+				   && *(p + 2) == 's'
+				   && *(p + 3) == 't'))
+			{
+				__FSM_H2_FIN(Req_HdrHostV, 4,
+					     TFW_TAG_HDR_HOST);
+			}
+			__FSM_H2_NEXT(Req_HdrH);
+		case 'i':
+			if (likely(__data_available(p, 17)
+				   && *(p + 1) == 'f'
+				   && *(p + 2) == '-'
+				   && C8_INT(p + 3, 'm', 'o', 'd', 'i',
+					     'f', 'i', 'e', 'd')
+				   && *(p + 11) == '-'
+				   && C4_INT(p + 12, 's', 'i', 'n', 'c')
+				   && *(p + 16) == 'e'))
+			{
+				__FSM_H2_FIN(Req_HdrIf_Modified_SinceV, 17,
+					     TFW_TAG_HDR_IF_MODIFIED_SINCE);
+			}
+			if (likely(__data_available(p, 13)
+				   && *(p + 1) == 'f'
+				   && *(p + 2) == '-'
+				   && C4_INT(p + 3, 'n', 'o', 'n', 'e')
+				   && *(p + 7) == '-'
+				   && C4_INT(p + 8, 'm', 'a', 't', 'c')
+				   && *(p + 12) == 'h'))
+			{
+				__FSM_H2_FIN(Req_HdrIf_None_MatchV, 13,
+					     TFW_TAG_HDR_IF_NONE_MATCH);
+			}
+			__FSM_H2_NEXT(Req_HdrI);
+		case 'k':
+			if (likely(__data_available(p, 10)
+				   && C4_INT(p, 'k', 'e', 'e', 'p')
+				   && *(p + 4) == '-'
+				   && C4_INT(p + 5, 'a', 'l', 'i', 'v')
+				   && *(p + 9) == 'e'))
```

Comment:
2 8-byte and 4-byte overlapping comparisons. And adjust all the other cases accordingly.